File: LeetCode/LeetCode 110. 平衡二叉树 (DFS).py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isBalanced(self, root: TreeNode) -> bool:

        def dfs(root: TreeNode) -> int:
            if root is None:
                return 0
            return max(dfs(root.left), dfs(root.right)) + 1
        
        if root is None:
            return True
        return abs(dfs(root.left) - dfs(root.right)) <= 1 and self.isBalanced(root.left) and self.isBalanced(root.right)

File: LeetCode/LeetCode 110. 平衡二叉树 (DFS).cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool ans = true;

    int dfs(TreeNode* root) {
        if(!root || !ans) return 0;
        int left_deep = dfs(root->left);
        int right_deep = dfs(root->right);
        if(abs(left_deep - right_deep) > 1) {
            ans = false;
            return 0;
        }
        return max(right_deep, left_deep) + 1;
    }

    bool isBalanced(TreeNode* root) {
        dfs(root);
        return ans;
    }
};